clang/include/clang/CConv/ConstraintsGraph.h

```diff
@@ -12,125 +12,269 @@
 #ifndef _CONSTRAINTSGRAPH_H
 #define _CONSTRAINTSGRAPH_H
 
-#include <boost/graph/graph_traits.hpp>
-#include <boost/graph/adjacency_list.hpp>
 #include "clang/CConv/Constraints.h"
+#include "llvm/ADT/DirectedGraph.h"
+#include "llvm/Support/GraphWriter.h"
+#include "llvm/ADT/GraphTraits.h"
+#include "llvm/ADT/BreadthFirstIterator.h"
 
-using namespace std;
+template<class DataType> struct DataEdge;
 
-template<class G>
-class BaseGraph {
+template<typename DataType, typename EdgeType = DataEdge<DataType>>
+class DataNode : public llvm::DGNode<DataNode<DataType, EdgeType>, EdgeType> {
 public:
-  typedef typename boost::graph_traits<G>::vertex_descriptor vertex_t;
-  typedef std::map<Atom*, vertex_t> VertexMapType;
-  G CG;
-protected:
-  VertexMapType AtomToVDMap;
+  typedef DataNode<DataType, EdgeType> NodeType;
+  typedef llvm::DGNode<NodeType, EdgeType> SuperType;
 
-  virtual vertex_t addVertex(Atom *A) {
-    if (AtomToVDMap.find(A) == AtomToVDMap.end()) {
-      auto Vidx = add_vertex(A, CG);
-      AtomToVDMap[A] = Vidx;
-    }
-    return AtomToVDMap[A];
+  DataNode() = delete;
+  DataNode(DataType D, EdgeType &E) : SuperType(E), Data(D) {}
+  explicit DataNode(DataType D) : SuperType(), Data(D) {}
+  DataNode(const DataNode &N) : SuperType(N), Data(N.Data) {}
+  DataNode(const DataNode &&N) : SuperType(std::move(N)), Data(N.Data) {}
+
+  ~DataNode() = default;
+
+  DataType getData() const { return Data; }
+
+  void addPred(EdgeType &E){ PredEdges.insert(&E); }
+  const llvm::SetVector<EdgeType *> &getPreds() { return PredEdges; }
+
+  bool isEqualTo(const DataNode &N) const { return this->Data == N.Data; }
```

Comment:
Why does equality ignore the `PredEdges` ?

Reply:
It's a resulet of an oddity of the llvm graph class. To check if a node is in the graph, you make an "empty" node (without any edges) but with whatever other properties you want to look for and call `findNode` on that. For this to work,  edges can't be part of the comparison.

Reply:
Add a comment to this effect?

Reply:
(Now I understand why you ignore the PredEdges -- not material to the identity of the node.)